### DISCUSSION TITLE
[ Terraform / modules ] Support for gitlab-tags

### DISCUSSION THREAD

**Discussion by xuoja-acc** [renovatebot/renovate#10943] [Request Help]
Using GitLab & Renovate self-hosted, we are trying to configure Renovate efficiently to scan our Terraform code and manage modules updates.

Terraform modules are also self-hosted on the same Gitlab instance and the RENOVATE_TOKEN give renovate all access needed to check tags/versions of the tf modules but the token is not used because terraform modules use git-tags as datasource : 
  ```
  TRACE: Cache miss (repository=terraform/infra/static_website)
       "namespace": "git-raw-refs",
       "key": "https://gitlab.example.net/terraform/infra/static_website.git"
DEBUG: Datasource unknown error (repository=terraform/infra/static_website)
       "datasource": "git-tags",
       "lookupName": "https://gitlab.example.net/terraform-modules/static-website.git",
       "err": {
         "task": {
           "commands": [
             "ls-remote",
             "https://gitlab.example.net/terraform-modules/static-website.git"
           ],
           "format": "utf-8"
         },
         "message": "fatal: could not read Username for 'https://gitlab.example.net': No such device or address\n",
         "stack": "Error: fatal: could not read Username for 'https://gitlab.example.net': No such device or address\n\n    at Object.action (/usr/src/app/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:38:28)\n    at PluginStore.exec (/usr/src/app/node_modules/simple-git/src/lib/plugins/plugin-store.ts:24:29)\n    at /usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:114:40\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:111:14)\n    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:88:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.js:5:58)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)"
       }
```

For now we are using [hostRules](https://docs.renovatebot.com/private-modules/#hostrules-configuration-using-environment-variables) as a workaround(see config.js below), but we'd like to know if there is a more efficient way ? Will it be possible to add gitlab-tags datasource support for terraform modules (and obviously use the RENOVATE_TOKEN to list tags)? 

<details>
  <summary>config.js</summary>

  ```
  module.exports = {
   requireConfig: false,
   onboarding: false,
   hostRules: [
      {
        hostType: 'git-tags',
        matchHost: process.env.CI_SERVER_HOST,
        username: 'nobody',
        password: process.env.RENOVATE_TOKEN,
      }
    ],
   extends: [
     "config:base",
   ]
  };
  ```
</details>

Best regards!

**Comment by viceice** [comment]
you can assign `gitlab-tags` datasource via package rule with `matchSourceUrls`

**Reply by xuoja-acc** [reply]
I don't see any reference of datasource setting in [the documentation](https://docs.renovatebot.com/configuration-options/) but I just tried this as you suggest  : 

```
  "packageRules": [
    {
      "matchSourceUrlPrefixes": ["https://gitlab.example.net/"],
      "datasource": "gitlab-tags"
    }
  ],
```
Still no effects :( : maybe i missed something.

According to the documentation and my understanding https://docs.renovatebot.com/modules/datasource/
`You do not need to ever configure/override datasources directly, but you may use them in a packageRules array to configure other aspects of Renovate's behavior, e.g.`

**Reply by rarkins** [reply]
You need to use something like matchPackageNames. Source URL is retrieved from the datasource.. so you can't do a rule to change the datasource based on a field which isn't known yet.

**Reply by xuoja-acc** [reply]
Thank's, i tried another way :
```
  "packageRules": [
    {
      "matchPackagePatterns": ["terraform-modules/"],
      "datasource": "gitlab-tags"
    },
```

This seems to effectively change the datasource git-tags to gitlab-tags ... but the constructed url is buggy (gitlab.com is added  at the url start and then the self hosted gitlab url is append) :
```
TRACE: got request (repository=terraform/infra/static_website)
       "url": "https://gitlab.com/api/v4/projects/https%3A%2F%2Fgitlab.example.net%2Fterraform-modules%2Fstatic-website.git/repository/tags?per_page=100",
       "options": {
         "method": "get",
         "hostType": "gitlab",
         "baseUrl": "https://gitlab.example.net/api/v4/",
         "paginate": true,
         "responseType": "json",
         "throwHttpErrors": true,
         "hooks": {"beforeRedirect": [null]},
         "headers": {
           "user-agent": "RenovateBot/25.56.9 (https://github.com/renovatebot/renovate)"
         },
         "timeout": 60000
       }
```

**Reply by rarkins** [reply]
I think git-tags is fully qualified while gitlab-tags expects just the project name. Possibly we could update gitlab-tags code to massage any full url

**Comment by Fuochi** [comment]
Hi @rarkins,
Sorry for writing in this old thread. Do you think it's possible to apply the same logic in terraform implemented by #27903 and #28075 for terragrunt?
